File: 10-heap/00295-Find_Median_From_Data_Stream/00295-find_median_from_data_stream.cpp
// 295: Find Median From Data Stream
// https://leetcode.com/problems/find-median-from-data-stream

#include <queue>
#include <vector>
#include <iostream>

using namespace std;


// SOLUTION
class MedianFinder {
private:
    priority_queue<int> lower;
    priority_queue<int, vector<int>, greater<int>> higher;

public:
    MedianFinder(){}
    
    void addNum(int num) {
        if (lower.empty()) {
            lower.push(num);
            return;
        }
        
        if (lower.size() > higher.size()) {
            if (lower.top() > num) {
                higher.push(lower.top());
                lower.pop();
                lower.push(num);
            } else {
                higher.push(num);
            }
        } else {
            if (num > higher.top()) {
                lower.push(higher.top());
                higher.pop();
                higher.push(num);
            } else {
                lower.push(num);
            }
        }
    }
    
    double findMedian() {
        double result = 0.0;
        
        if (lower.size() == higher.size()) {
            result = lower.top() + (higher.top() - lower.top()) / 2.0;
        } else {
            if (lower.size() > higher.size()) {
                result = lower.top();
            } else {
                result = higher.top();
            }
        }
        
        return result;
    }
};


int main() {
    MedianFinder *o = new MedianFinder();

    // OPERATIONS
    o->addNum(1);
    o->addNum(2);
    cout << o->findMedian() << endl;
    o->addNum(3);
    cout << o->findMedian() << endl;

    return 0;
}

File: 10-heap/00295-Find_Median_From_Data_Stream/00295-find_median_from_data_stream.py
# 295: Find Median From Data Stream
# https://leetcode.com/problems/find-median-from-data-stream

import heapq

# SOLUTION
class MedianFinder:
    def __init__(self):
        self.lower, self.higher = [], []

    def addNum(self, num: int) -> None:
        if self.higher and num > self.higher[0]:
            heapq.heappush(self.higher, num)
        else:
            heapq.heappush(self.lower, -1 * num)

        if len(self.lower) > len(self.higher) + 1:
            value = -1 * heapq.heappop(self.lower)
            heapq.heappush(self.higher, value)

        if len(self.higher) > len(self.lower) + 1:
            value = heapq.heappop(self.higher)
            heapq.heappush(self.lower, -1 * value)

    def findMedian(self) -> float:
        if len(self.lower) > len(self.higher):
            return -1 * self.lower[0]
        elif len(self.higher) > len(self.lower):
            return self.higher[0]
        return (-1 * self.lower[0] + self.higher[0]) / 2.0


if __name__ == "__main__":
    o = MedianFinder()

    # OPERATIONS
    o.addNum(1)
    o.addNum(2)
    print(o.findMedian())
    o.addNum(3)
    print(o.findMedian())
